File: 81 (Search in Rotated Sorted Array II)/cpp/Solution.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size()-1;

        while (low<=high){
            while (low<high && nums[low]==nums[low+1]) low++;
            while (low<high && nums[high]== nums[high-1]) high--;


            int mid = (low+high)>>1;

            if (nums[mid]==target) return true;

            if (nums[low]<=nums[mid]){
                if (nums[low]<=target && nums[mid]>=target) high = mid-1;
                else low = mid+1;
            }
            else{
                if (nums[mid]<=target && nums[high]>=target) low = mid+1;
                else high = mid-1;
            }
        }
        return false;
    }
};

File: 81 (Search in Rotated Sorted Array II)/python/Solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        low = 0
        high = len(nums)-1

        while low<=high:

            while low<high and nums[low]==nums[low+1]: low+=1
            while low<high and nums[high]==nums[high-1]: high-=1

            mid = (low+high)>>1


            if (nums[mid]==target): return True

            if nums[low]<=nums[mid]:
                if (nums[low]<=target and nums[mid]>=target): high = mid-1
                else: low = mid+1
            
            else:
                if (nums[mid]<=target and nums[high]>=target): low = mid+1
                else: high = mid-1
            
        
        return False